File: Assets/Mediapipe/SDK/Scripts/ImageFrame.cs
using System;
using System.Runtime.InteropServices;
using Unity.Collections;
using Unity.Collections.LowLevel.Unsafe;
using UnityEngine;

using ImageFramePtr = System.IntPtr;

namespace Mediapipe {
  public class ImageFrame : ResourceHandle {
    /**
    * Constants (TODO: read from ddl)
    */
    public static readonly uint kDefaultAlignmentBoundary = 16;
    public static readonly uint kGlDefaultAlignmentBoundary = 4;

    private bool _disposed = false;
    private GCHandle freePixelDataHandle;

    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
    private delegate void FreeMemoryHandler(IntPtr ptr);

    public ImageFrame() : base(UnsafeNativeMethods.MpImageFrameCreateDefault()) {}

    public ImageFrame(ImageFramePtr imageFramePtr, bool isOwner = true) : base(imageFramePtr, isOwner) {}

    public ImageFrame(ImageFormat format, int width, int height, uint alignmentBoundary) :
      base(UnsafeNativeMethods.MpImageFrameCreate((int)format, width, height, alignmentBoundary)) {}

    public ImageFrame(ImageFormat format, int width, int height, int widthStep, NativeArray<byte> pixelData) {
      FreeMemoryHandler memoryHandler = (IntPtr ptr) => { pixelData.Dispose(); };
      freePixelDataHandle = GCHandle.Alloc(memoryHandler, GCHandleType.Pinned);

      unsafe {
        ptr = UnsafeNativeMethods.MpImageFrameCreateWithPixelData(
          (int)format, width, height, widthStep,
          (IntPtr)NativeArrayUnsafeUtility.GetUnsafeReadOnlyPtr(pixelData),
          Marshal.GetFunctionPointerForDelegate(memoryHandler)
        );
      }

      base.TakeOwnership(ptr);
    }

    protected override void Dispose(bool disposing) {
      if (_disposed) return;

      if (OwnsResource()) {
        // NOTE: Below line will call FreePixelData if neccessary.
        UnsafeNativeMethods.MpImageFrameDestroy(ptr);
      }

      if (freePixelDataHandle != null && freePixelDataHandle.IsAllocated) {
        freePixelDataHandle.Free();
      }

      ptr = IntPtr.Zero;

      _disposed = true;
    }

    public ImageFormat Format() {
      return (ImageFormat)UnsafeNativeMethods.MpImageFrameFormat(ptr);
    }

    public int Width() {
      return UnsafeNativeMethods.MpImageFrameWidth(ptr);
    }

    public int Height() {
      return UnsafeNativeMethods.MpImageFrameHeight(ptr);
    }

    public int ChannelSize() {
      return UnsafeNativeMethods.MpImageFrameChannelSize(ptr);
    }

    public int NumberOfChannels() {
      return UnsafeNativeMethods.MpImageFrameNumberOfChannels(ptr);
    }

    public int ByteDepth() {
      return UnsafeNativeMethods.MpImageFrameByteDepth(ptr);
    }

    public int WidthStep() {
      return UnsafeNativeMethods.MpImageFrameWidthStep(ptr);
    }

    public Color32[] GetColor32s() {
      return Mediapipe.Format.FromBytePtr(PixelDataPtr(), Format(), Width(), Height(), WidthStep());
    }

    public IntPtr PixelDataPtr() {
      return UnsafeNativeMethods.MpImageFramePixelData(ptr);
    }

    public static unsafe ImageFrame FromPixels32(Color32[] colors, int width, int height) {
      return new ImageFrame(ImageFormat.SRGB, width, height, 3 * width, Mediapipe.Format.FromPixels32(colors, width, height));
    }
  }
}

File: C/mediapipe_api/framework/formats/image_frame.cc
#include "mediapipe_api/framework/formats/image_frame.h"

mediapipe::ImageFrame* MpImageFrameCreate(int format_code, int width, int height, uint32 alignment_boundary) {
  auto format = static_cast<mediapipe::ImageFormat::Format>(format_code);

  return new mediapipe::ImageFrame { format, width, height, alignment_boundary };
}

mediapipe::ImageFrame* MpImageFrameCreateDefault() {
  return new mediapipe::ImageFrame();
}

mediapipe::ImageFrame* MpImageFrameCreateWithPixelData(
    int format_code, int width, int height, int width_step, uint8* pixel_data, Deleter deleter) {
  auto format = static_cast<mediapipe::ImageFormat::Format>(format_code);

  return new mediapipe::ImageFrame { format, width, height, width_step, pixel_data, deleter };
}

void MpImageFrameDestroy(mediapipe::ImageFrame* image_frame) {
  return delete image_frame;
}

bool MpImageFrameIsEmpty(mediapipe::ImageFrame* image_frame) {
  return image_frame->IsEmpty();
}

int MpImageFrameFormat(mediapipe::ImageFrame* image_frame) {
  return static_cast<int>(image_frame->Format());
}

int MpImageFrameWidth(mediapipe::ImageFrame* image_frame) {
  return image_frame->Width();
}

int MpImageFrameHeight(mediapipe::ImageFrame* image_frame) {
  return image_frame->Height();
}

int MpImageFrameChannelSize(mediapipe::ImageFrame* image_frame) {
  return image_frame->ChannelSize();
}

int MpImageFrameNumberOfChannels(mediapipe::ImageFrame* image_frame) {
  return image_frame->NumberOfChannels();
}

int MpImageFrameByteDepth(mediapipe::ImageFrame* image_frame) {
  return image_frame->ByteDepth();
}

int MpImageFrameWidthStep(mediapipe::ImageFrame* image_frame) {
  return image_frame->WidthStep();
}

uint8* MpImageFramePixelData(mediapipe::ImageFrame* image_frame) {
  return image_frame->MutablePixelData();
}

void MpStatusOrImageFrameDestroy(MpStatusOrImageFrame* status_or_image_frame) {
  delete status_or_image_frame;
}

MpStatus* MpStatusOrImageFrameStatus(MpStatusOrImageFrame* status_or_image_frame) {
  mediapipe::Status status { *status_or_image_frame->status };

  return new MpStatus { std::move(status) };
}

mediapipe::ImageFrame* MpStatusOrImageFrameConsumeValue(MpStatusOrImageFrame* status_or_image_frame) {
  return status_or_image_frame->value.release();
}

MpPacket* MpMakeImageFramePacketAt(mediapipe::ImageFrame* image_frame, int timestamp) {
  auto packet = mediapipe::Adopt(image_frame).At(mediapipe::Timestamp(timestamp));

  return new MpPacket { std::move(packet) };
}

mediapipe::ImageFrame* MpPacketGetImageFrame(MpPacket* packet) {
  auto holder = static_cast<const UnsafePacketHolder<mediapipe::ImageFrame>*>(mediapipe::packet_internal::GetHolder(*packet->impl));

  return holder->Get();
}

MpStatusOrImageFrame* MpPacketConsumeImageFrame(MpPacket* packet) {
  auto status_or_image_frame = packet->impl->Consume<mediapipe::ImageFrame>();

  return new MpStatusOrImageFrame { std::move(status_or_image_frame) };
}
